### PR TITLE
PERF: vectorized DateOffset with months

### PR DESCRIPTION
**Pull request by chris-b1**
This is a follow-up to https://github.com/pydata/pandas/pull/10744.  In that, vectorized versions of some offsets were implemented, mostly by changing to periods and back.  

The case of shifting by years/months (which is actually most useful to me) required some extra hoops and had poorer performance - this PR implements a special cython routine for that, for about a 10x improvement. 

```
In [3]: s = pd.Series(pd.date_range('1900-1-1', periods=100000))

# Master
In [4]: %timeit s + pd.DateOffset(months=1)
1 loops, best of 3: 140 ms per loop

# PR
In [2]: %timeit s + pd.DateOffset(months=1)
100 loops, best of 3: 14.2 ms per loo
```

### CODE REVIEW

File: pandas/tseries/offsets.py [hunk 1]
```diff
@@ -261,16 +261,12 @@ def apply_index(self, i):
...
             if months:
-                base = (i.to_period('M') + months).to_timestamp()
-                time = i.to_perioddelta('D')
-                days = i.to_perioddelta('M') - time
-                # minimum prevents month-end from wrapping
-                day_offset = np.minimum(days,
-                                        to_timedelta(base.days_in_month - 1, unit='D'))
-                i = base + day_offset + time
+                shifted = tslib.shift_months(i.asi8, months)
+                i = i._constructor(shifted)
```
Comment (maximilianr) [inline pandas/tseries/offsets.py]: This will lose the `name` of the `Index` - I think you want to use `_shallow_copy`:

```python

In [57]: df.index._constructor(df.index)
Out[57]: 
Int64Index([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18,
            19],
           dtype='int64')

In [58]: df.index._shallow_copy(df.index)
Out[58]: 
Int64Index([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18,
            19],
           dtype='int64', name='hi')
```
(there's a broader issue about the constructor keeping the name)

Comment (chris-b1) [inline pandas/tseries/offsets.py]: It actually doesn't cause a problem in this case because the values are going to be unboxed/boxed here anyways:
https://github.com/pydata/pandas/blob/master/pandas/tseries/index.py#L716

But I'll change it if it's the more idiomatic way to get the constructor?

Comment (jreback) [inline pandas/tseries/offsets.py]: yes shallow copy is the idiom

File: pandas/tseries/tests/test_tslib.py [hunk 2]
```diff
@@ -949,6 +949,17 @@ def compare_local_to_utc(tz_didx, utc_didx):
                                   tslib.maybe_get_tz('Asia/Tokyo'))
         self.assert_numpy_array_equal(result, np.array([tslib.iNaT], dtype=np.int64))
 
+    def test_shift_months(self):
+        s = DatetimeIndex([Timestamp('2000-01-05 00:15:00'), Timestamp('2000-01-31 00:23:00'),
+                           Timestamp('2000-01-01'), Timestamp('2000-02-29'), Timestamp('2000-12-31')])
+        for years in [-1, 0, 1]:
```
Comment (jreback) [inline pandas/tseries/tests/test_tslib.py]: add a name in the test

File: pandas/tslib.pyx [hunk 3]
```diff
@@ -4386,6 +4387,40 @@ cpdef normalize_date(object dt):
...
+    specified number of months using relativedelta semantics
+    '''
+    cdef:
+        Py_ssize_t i
+        int isleap, days_in_month, new_month
+        pandas_datetimestruct dts
+        int count = len(dtindex)
+        int64_t[:] out = np.empty(count, dtype='int64')
+
+    count = len(dtindex)
```
Comment (jreback) [inline pandas/tslib.pyx]: you defined count, out twice

File: pandas/tslib.pyx [hunk 4]
```diff
@@ -4386,6 +4387,40 @@ cpdef normalize_date(object dt):
...
+    out = np.empty(count, dtype='int64')
+
+    for i in range(count):
+        if dtindex[i] == NPY_NAT:
+            out[i] = NPY_NAT
+        else:
+            pandas_datetime_to_datetimestruct(dtindex[i], PANDAS_FR_ns, &dts)
```
Comment (jreback) [inline pandas/tslib.pyx]: this routine could actually be used in the Dateoffset for adding months (possibly dealing with month end/start as well)

could do in another PR though

**Comment by chris-b1** [comment]
@jreback - pushed changes for your comments and added `MonthEnd` and `MonthBegin`

For your comment about using this routine in the `DateOffset` are you talking about cythonizing the actual `apply` method too?  On that front, I see there is an `offsets.pyx` file, but doesn't look it's actually used?

File: pandas/tseries/tests/test_tslib.py (hunk 2 above)
Comment (chris-b1) [inline pandas/tseries/tests/test_tslib.py]: didn't add here because the `tslib` method just operates on the ints, but I did add names to some of the existing `DateOffset` tests

File: pandas/tslib.pyx [hunk 5]
```diff
@@ -4386,6 +4387,73 @@ cpdef normalize_date(object dt):
         raise TypeError('Unrecognized type: %s' % type(dt))
 
 
+cdef inline int _year_add_months(pandas_datetimestruct dts,
+                                 int months):
```
Comment (jreback) [inline pandas/tslib.pyx]: can you add a doc-string to these

**Comment by jreback** [comment]
do we have an asv benchmark for this? if not pls add one. otherwise minor change. lmk

**Comment by chris-b1** [comment]
Made those doc changes.  Yeah,  there is an `asv` for `DateOffset`

        before     after       ratio
      [8ea8968 ] [c1d53f9 ]
    -  121.83ms    18.49ms      0.15 timeseries.timeseries_datetimeindex_offset_fast.time_timeseries_datetimeindex_offset_fast
    -  121.50ms    21.70ms      0.18  timeseries.timeseries_series_offset_fast.time_timeseries_series_offset_fast

**Comment by jreback** [comment]
@chris-b1 thanks! these pr's are awesome! keep em coming!
